### PR TITLE
Ability to create local RenderingDevice instances.

### PR DESCRIPTION
**Pull request by reduz**
You can do
```
var rd = RenderingDevice.create_local_device()
# use rd to do all local drawing, compute, etc on any thread you want
rd.free()
```
that's it.

### CODE REVIEW

**Comment by reduz** [comment]
These RenderingDevices can't draw to the screen, nor share data with the ones used for rendering. They are mostly meant so you can take advantage of the GPU for complex game logic (very useful on sandbox style games).

File: drivers/vulkan/vulkan_context.cpp [hunk 1]
```diff
@@ -1499,6 +1499,87 @@ VulkanContext::VulkanContext() {
...
+			.queueCreateInfoCount = 1,
+			.pQueueCreateInfos = queues,
+			.enabledLayerCount = 0,
+			.ppEnabledLayerNames = nullptr,
+			.enabledExtensionCount = enabled_extension_count,
+			.ppEnabledExtensionNames = (const char *const *)extension_names,
+			.pEnabledFeatures = &physical_device_features, // If specific features are required, pass them in here
+		};
+		err = vkCreateDevice(gpu, &sdevice, nullptr, &ld.device);
+		ERR_FAIL_COND_V(err, RID());
```
Comment (clayjohn) [inline drivers/vulkan/vulkan_context.cpp:1529]: We should probably provide a more robust error here. Many GPUs are limited to a single graphics queue (most AMD and all INTEL AFAIK) and will error out here when attempting to make a second graphics queue.

Comment (clayjohn) [reply drivers/vulkan/vulkan_context.cpp:1529]: As per our IRC conversation. This was my bad. I mixed up logical devices and physical devices.

**Review by clayjohn — APPROVED** [review]

File: drivers/vulkan/vulkan_context.cpp [hunk 2]
```diff
@@ -1499,6 +1499,87 @@ VulkanContext::VulkanContext() {
...
+		VkDeviceQueueCreateInfo queues[2];
+		queues[0].sType = VK_STRUCTURE_TYPE_DEVICE_QUEUE_CREATE_INFO;
+		queues[0].pNext = nullptr;
+		queues[0].queueFamilyIndex = graphics_queue_family_index;
+		queues[0].queueCount = 1;
+		queues[0].pQueuePriorities = queue_priorities;
+		queues[0].flags = 0;
+
+		VkDeviceCreateInfo sdevice = {
+			.sType = VK_STRUCTURE_TYPE_DEVICE_CREATE_INFO,
```
Comment (clayjohn) [inline drivers/vulkan/vulkan_context.cpp:1517]: This is where Appveyor is erroring out.

**Comment by akien-mga** [comment]
Thanks!

**Comment by neikeq** [comment]
This broke the C# bindings. I'm confused with the decision to make `RenderingDevice` a singleton class but also not a singleton class :confused:

**Comment by akien-mga** [comment]
We discussed further on IRC but for the sake of consistency: `RenderingDevice` was made a singleton by mistake in a follow-up PR: #38431. It's not clear how the API can be used at all from script though.

**Comment by Tuatarian** [comment]
This is effectively allowing us to do compute shaders from GDscript right? Looking forward

**Comment by Calinou** [comment]
@Tuatarian Yes, but there's no documentation on it yet. I guess the API will need to change a bit to allow for easier usability as well. Either way, it's still considered an advanced use case.

**Comment by Tuatarian** [comment]
@Calinou I don't know how "advanced", compute shaders necessarily are. Not to familiar with this API, hope we get some docs before release

**Pull request merged**